src/v/datalake/record_schema_resolver.cc

```diff
@@ -0,0 +1,206 @@
+              std::move(parsable_buf));
+            auto decoder = avro::validatingDecoder(s, avro::binaryDecoder());
+            decoder->init(*in);
+            avro::decode(*decoder, d);
+        } catch (...) {
+            co_return value_conversion_exception(fmt::format(
+              "Error reading Avro buffer: {}", std::current_exception()));
+        }
+        co_return iceberg::val_from_avro(d, type, iceberg::field_required::yes);
+    }
+};
+
+namespace ppsr = pandaproxy::schema_registry;
+struct schema_translating_visitor {
+    enum class errc {
+        invalid_buf,
+        schema_translation_error,
+        not_supported,
+    };
+    // Buffer without the schema ID.
+    iobuf buf_no_id;
+
+    ss::future<checked<type_and_buf, errc>>
+    operator()(const ppsr::avro_schema_definition& avro_def) {
+        const auto& avro_schema = avro_def();
+        try {
+            // TODO: we should use Avro value conversion instead of this
+            // Iceberg-specific code.
+            auto type = iceberg::type_from_avro(
+              avro_schema.root(), iceberg::with_field_ids::no);
+            co_return type_and_buf{
+              .schema = avro_schema,
+              .pb_offsets = std::nullopt,
+              .type = std::move(type),
+              .parsable_buf = std::move(buf_no_id),
+            };
+        } catch (...) {
+            vlog(
+              datalake_log.error,
+              "Avro schema translation failed: {}",
+              std::current_exception());
+            co_return errc::schema_translation_error;
+        }
+    }
+    ss::future<checked<type_and_buf, errc>>
+    operator()(const ppsr::protobuf_schema_definition&) {
+        // TODO: will be implemented in a subsequent commit!
+        co_return errc::not_supported;
+    }
+    ss::future<checked<type_and_buf, errc>>
+    operator()(const ppsr::json_schema_definition&) {
+        co_return errc::not_supported;
+    }
+};
+
+} // namespace
+
+resolved_buf resolved_buf::make_raw_binary(iobuf b) {
+    return resolved_buf{
+      .schema_identifier = std::nullopt,
+      .type = std::nullopt,
+      .val = binary_val(std::move(b)),
+    };
+}
+
+ss::future<checked<resolved_buf, record_schema_resolver::errc>>
+record_schema_resolver::resolve_buf_schema(iobuf b) const {
+    schema_message_data schema_id_res;
+    try {
+        // NOTE: Kafka's serialization protocol relies on a magic byte to
+        // indicate if we have a schema. This has room for false positives, and
+        // we can't say for sure if an error is the result of the record not
+        // having a schema. Just translate to binary.
+        auto res = get_value_schema_id(b);
+        if (res.has_error()) {
+            co_return resolved_buf::make_raw_binary(std::move(b));
+        }
+        schema_id_res = std::move(res.value());
+    } catch (...) {
+        co_return resolved_buf::make_raw_binary(std::move(b));
+    }
```

Comment:
consider adding logging here if resolving schema failed

Reply:
It seems like this here is possible to get here when there is no schema in the data. This shouldn't be an error and I worry may be a bit verbose. In any case, I added a trace level log, and we can reduce logging as needed.

Reply:
Done